lib/rubygems.rb

```diff
@@ -1236,20 +1238,40 @@ class << self
   end
 
   ##
-  # The SOURCE_DATE_EPOCH environment variable (or, if that's not set, the current time), converted to Time object.
-  # This is used throughout RubyGems for enabling reproducible builds.
-  #
-  # If it is not set as an environment variable already, this also sets it.
+  # If the SOURCE_DATE_EPOCH environment variable, returns it's value.
+  # Otherwise, returns the current time in a format suitable for that variable.
   #
   # Details on SOURCE_DATE_EPOCH:
   # https://reproducible-builds.org/specs/source-date-epoch/
 
+  def self.raw_source_date_epoch
+    epoch = ENV["SOURCE_DATE_EPOCH"]
+    default_epoch = Time.now.to_i.to_s
+
+    return epoch unless epoch.nil? || epoch.strip.empty?
+
+    default_epoch
+  end
+
+  ##
+  # Returns the value of Gem.raw_source_date_epoch, as a Time object.
+  #
+  # This is used throughout RubyGems for enabling reproducible builds.
   def self.source_date_epoch
-    if ENV["SOURCE_DATE_EPOCH"].nil? || ENV["SOURCE_DATE_EPOCH"].empty?
-      ENV["SOURCE_DATE_EPOCH"] = Time.now.to_i.to_s
-    end
+    Time.at(self.raw_source_date_epoch.strip.to_i).utc.freeze
+  end
 
-    Time.at(ENV["SOURCE_DATE_EPOCH"].to_i).utc.freeze
+  ##
+  # Sets the SOURCE_DATE_EPOCH environment variable, runs +block+,
+  # resets SOURCE_DATE_EPOCH, and then returns the result from running +block+.
+  def self.with_source_date_epoch(&block)
+    old_epoch = ENV["SOURCE_DATE_EPOCH"]
+    new_epoch = self.raw_source_date_epoch
+
+    ENV["SOURCE_DATE_EPOCH"] = new_epoch unless old_epoch == new_epoch
```

Comment:
It is necessary that we still set this environment variable? could we manage globally the state of the build time internally(using `SOURCE_DATE_EPOCH`) in Rubygems instead?

Reply:
So, I'm not entirely sure. For now, I settled for setting it only when needed. As far as I know, the only way it would be a problem is if it spawns any subprocesses. Whenever RubyGems spawns processes where we don't control all of the code, like if it's building a C extension, then it's completely unavoidable.

This is partially related to what I mentioned above, about wanting tests for reproducible builds + C extensions and such.

Reply:
I basically approached this first attempt by tightening up the timeframe during which we set the environment variable. If it can be tightened even farther, I would definitely like to. But I think reproducible builds for C extensions would be a prerequisite. (Well, assuming that even works right now? I'm not sure if it does.)

Reply:
Also: I _think_ the places I use `Gem.with_source_date_epoch { ... }` are only called during the `build`/`install` subcommands, but I've not confirmed this.

Reply:
Actually, I just missed something very obvious &mdash; if we can find any places where subprocesses would be started during `gem build`/`rake build` and the like, I think we could just put it into the subprocesses environment and avoid it entirely! So, just need to find out if and where that happens, and find a way to tie it in. :slightly_smiling_face: